iaputil/iap_jwt_auth.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iaputil
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/abcxyz/pkg/jwtutil"
+)
+
+// IAPJWTAuthenticationHandler allows for retrieving principal information from JWT tokens provided by
+// Identity Aware Proxy (IAP).
+type IAPJWTAuthenticationHandler struct {
+	*jwtutil.Verifier
+	// PrincipalClaimKey is the key in the JWTs claims which corresponds to the user's email.
+	PrincipalClaimKey string
+	// Endpoint is the endpoint where JWKs public keys can be found to do JWT validation.
+	Endpoint string
+}
+
+// NewIAPJWTAuthenticationHandler returns a IAPJWTAuthenticationHandler with a verifier initialized. Uses defaults
+// for JWT related fields that will retrieve a user email when using IAP on GCP.
+func NewIAPJWTAuthenticationHandler(ctx context.Context) (*IAPJWTAuthenticationHandler, error) {
+	j := &IAPJWTAuthenticationHandler{
+		PrincipalClaimKey: "email",
+		// For context: https://cloud.google.com/iap/docs/signed-headers-howto#verifying_the_jwt_header
+		Endpoint: "https://www.gstatic.com/iap/verify/public_key-jwk",
```

Comment:
The IAP docs have their own [jwt validator](https://github.com/googleapis/google-api-go-client/blob/b35900aff148bb8446ca7e15ed0beb50d749d6c7/idtoken/validate.go#L125-L165) which also validates the aud if/when we choose to validate that. 

If it doesn't matter I continued to leverage the one we are using (github.com/lestrrat-go/jwx/v2/jwt) for the IAP case. Idk if we need to find the public key associated to the kid of the JWT

From the IAP docs: 
https://www.gstatic.com/iap/verify/public_key. This URL contains a JSON dictionary that maps the kid claims to the public key values.
https://www.gstatic.com/iap/verify/public_key-jwk. This URL contains the IAP public keys in [JWK](https://tools.ietf.org/html/rfc7517) format.